File: cross-platform/windows-controller/src-tauri/src/adb.rs
use crate::{run_command, AdbDevice};
use base64::Engine;

/// List all ADB-connected devices (physical and emulator instances).
pub async fn list_devices(adb: &str) -> Result<Vec<AdbDevice>, String> {
    let output = run_command(adb, &["devices", "-l"]).await?;
    let mut devices = Vec::new();

    for line in output.lines().skip(1) {
        let line = line.trim();
        if line.is_empty() || line.starts_with('*') {
            continue;
        }

        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 2 {
            continue;
        }

        let serial = parts[0].to_string();
        let status = parts[1].to_string();

        // Parse key:value pairs from the rest
        let mut model = String::new();
        let mut product = String::new();
        let mut transport_id = String::new();

        for part in &parts[2..] {
            if let Some((key, value)) = part.split_once(':') {
                match key {
                    "model" => model = value.to_string(),
                    "product" => product = value.to_string(),
                    "transport_id" => transport_id = value.to_string(),
                    _ => {}
                }
            }
        }

        let is_emulator = serial.starts_with("emulator-");

        // Get Android version and API level if device is online
        let (android_version, api_level) = if status == "device" {
            let ver = get_prop(adb, &serial, "ro.build.version.release").await.unwrap_or_default();
            let api = get_prop(adb, &serial, "ro.build.version.sdk").await
                .unwrap_or_default()
                .trim()
                .parse::<u32>()
                .unwrap_or(0);
            (ver.trim().to_string(), api)
        } else {
            (String::new(), 0)
        };

        devices.push(AdbDevice {
            serial,
            device_type: status,
            model,
            product,
            transport_id,
            android_version,
            api_level,
            is_emulator,
        });
    }

    Ok(devices)
}

/// Get detailed device properties.
pub async fn get_device_details(adb: &str, serial: &str) -> Result<serde_json::Value, String> {
    let props = vec![
        ("manufacturer", "ro.product.manufacturer"),
        ("model", "ro.product.model"),
        ("brand", "ro.product.brand"),
        ("android_version", "ro.build.version.release"),
        ("api_level", "ro.build.version.sdk"),
        ("screen_density", "ro.sf.lcd_density"),
        ("abi", "ro.product.cpu.abi"),
        ("build_type", "ro.build.type"),
        ("build_display", "ro.build.display.id"),
        ("security_patch", "ro.build.version.security_patch"),
    ];

    let mut details = serde_json::Map::new();
    details.insert("serial".into(), serde_json::Value::String(serial.to_string()));

    for (key, prop) in props {
        let value = get_prop(adb, serial, prop).await.unwrap_or_default().trim().to_string();
        details.insert(key.into(), serde_json::Value::String(value));
    }

    // Get screen resolution
    let wm_output = run_command(adb, &["-s", serial, "shell", "wm", "size"]).await.unwrap_or_default();
    let resolution = wm_output
        .lines()
        .find(|l| l.contains("Physical size"))
        .map(|l| l.split(':').nth(1).unwrap_or("").trim().to_string())
        .unwrap_or_default();
    details.insert("screen_resolution".into(), serde_json::Value::String(resolution));

    // Get battery level
    let battery_output = run_command(adb, &["-s", serial, "shell", "dumpsys", "battery"]).await.unwrap_or_default();
    let battery_level = battery_output
        .lines()
        .find(|l| l.trim().starts_with("level:"))
        .map(|l| l.split(':').nth(1).unwrap_or("0").trim().to_string())
        .unwrap_or_else(|| "0".into());
    details.insert("battery_level".into(), serde_json::Value::String(battery_level));

    Ok(serde_json::Value::Object(details))
}

/// Take a screenshot and return base64-encoded PNG.
pub async fn take_screenshot(adb: &str, serial: &str) -> Result<String, String> {
    // screencap on device, pull to stdout as PNG
    let output = tokio::process::Command::new(adb)
        .args(["-s", serial, "exec-out", "screencap", "-p"])
        .output()
        .await
        .map_err(|e| format!("Failed to take screenshot: {}", e))?;

    if !output.status.success() {
        return Err(format!("screencap failed: {}", String::from_utf8_lossy(&output.stderr)));
    }

    let b64 = base64::engine::general_purpose::STANDARD.encode(&output.stdout);
    Ok(b64)
}

/// Send input events to a device via adb shell input.
pub async fn device_input(
    adb: &str,
    serial: &str,
    action: &str,
    x: Option<i32>,
    y: Option<i32>,
    to_x: Option<i32>,
    to_y: Option<i32>,
    text: Option<String>,
    key_code: Option<i32>,
    duration: Option<i32>,
) -> Result<String, String> {
    let args: Vec<String> = match action {
        "tap" => {
            let x = x.ok_or("tap requires x")?;
            let y = y.ok_or("tap requires y")?;
            vec!["-s".into(), serial.into(), "shell".into(), "input".into(), "tap".into(),
                 x.to_string(), y.to_string()]
        }
        "swipe" => {
            let x = x.ok_or("swipe requires x")?;
            let y = y.ok_or("swipe requires y")?;
            let tx = to_x.ok_or("swipe requires toX")?;
            let ty = to_y.ok_or("swipe requires toY")?;
            let dur = duration.unwrap_or(300);
            vec!["-s".into(), serial.into(), "shell".into(), "input".into(), "swipe".into(),
                 x.to_string(), y.to_string(), tx.to_string(), ty.to_string(), dur.to_string()]
        }
        "longPress" => {
            let x = x.ok_or("longPress requires x")?;
            let y = y.ok_or("longPress requires y")?;
            let dur = duration.unwrap_or(1000);
            // long press = swipe to same point with duration
            vec!["-s".into(), serial.into(), "shell".into(), "input".into(), "swipe".into(),
                 x.to_string(), y.to_string(), x.to_string(), y.to_string(), dur.to_string()]
        }
        "inputText" => {
            let txt = text.ok_or("inputText requires text")?;
            // Escape spaces for adb shell input text
            let escaped = txt.replace(' ', "%s");
            vec!["-s".into(), serial.into(), "shell".into(), "input".into(), "text".into(), escaped]
        }
        "keyEvent" => {
            let code = key_code.ok_or("keyEvent requires keyCode")?;
            vec!["-s".into(), serial.into(), "shell".into(), "input".into(), "keyevent".into(),
                 code.to_string()]
        }
        "dumpUi" => {
            // Dump UI hierarchy
            run_command(adb, &["-s", serial, "shell", "uiautomator", "dump", "/dev/tty"]).await?;
            return Ok("UI dump captured".into());
        }
        _ => return Err(format!("Unknown action: {}", action)),
    };

    let args_refs: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
    run_command(adb, &args_refs).await
}

/// Install an APK on a device.
pub async fn install_apk(adb: &str, serial: &str, apk_path: &str, reinstall: bool) -> Result<String, String> {
    let mut args = vec!["-s", serial, "install"];
    if reinstall {
        args.push("-r");
    }
    args.push("-g"); // grant all permissions
    args.push(apk_path);
    run_command(adb, &args).await
}

/// Uninstall a package from a device.
pub async fn uninstall_package(adb: &str, serial: &str, package_name: &str) -> Result<String, String> {
    run_command(adb, &["-s", serial, "uninstall", package_name]).await
}

/// List installed packages on a device.
pub async fn list_packages(adb: &str, serial: &str) -> Result<Vec<String>, String> {
    let output = run_command(adb, &["-s", serial, "shell", "pm", "list", "packages", "-3"]).await?;
    let packages: Vec<String> = output
        .lines()
        .filter_map(|line| line.strip_prefix("package:"))
        .map(|s| s.trim().to_string())
        .collect();
    Ok(packages)
}

/// Get a device property via adb shell getprop.
async fn get_prop(adb: &str, serial: &str, prop: &str) -> Result<String, String> {
    run_command(adb, &["-s", serial, "shell", "getprop", prop]).await
}

/// Get recent logcat output.
pub async fn get_logcat(adb: &str, serial: &str, lines: u32) -> Result<Vec<String>, String> {
    let output = run_command(adb, &[
        "-s", serial, "logcat", "-d", "-t", &lines.to_string(), "-v", "threadtime"
    ]).await?;
    Ok(output.lines().map(|l| l.to_string()).collect())
}

/// Clear logcat buffer.
pub async fn clear_logcat(adb: &str, serial: &str) -> Result<(), String> {
    run_command(adb, &["-s", serial, "logcat", "-c"]).await?;
    Ok(())
}

File: cross-platform/windows-controller/src-electron/services/adb.ts
// ADB service — wraps Android Debug Bridge CLI commands
// Ported from src-tauri/src/adb.rs

import { execFile } from "child_process";
import { promisify } from "util";
import type { AdbDevice, AdbDeviceDetails } from "./types";

const execFileAsync = promisify(execFile);

const TIMEOUT = 15_000;
const MAX_BUFFER = 50 * 1024 * 1024;

type ExecError = Error & {
  code?: string | number;
  cmd?: string;
  stderr?: string | Buffer;
  stdout?: string | Buffer;
};

const TRANSIENT_DAEMON_ERROR_PATTERNS = [
  /daemon not running/i,
  /cannot connect to daemon/i,
  /failed to start daemon/i,
  /failed to check server version/i,
  /adb server.*out of date/i,
  /cannot bind.*5037/i,
];

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function toText(value: unknown): string {
  if (Buffer.isBuffer(value)) return value.toString("utf-8");
  return typeof value === "string" ? value : String(value ?? "");
}

function errorText(error: unknown): string {
  const err = error as ExecError;
  return [
    err.message,
    toText(err.stderr),
    toText(err.stdout),
  ]
    .filter(Boolean)
    .join("\n");
}

function isTransientDaemonError(error: unknown): boolean {
  const text = errorText(error);
  return TRANSIENT_DAEMON_ERROR_PATTERNS.some((re) => re.test(text));
}

function formatAdbError(adbPath: string, args: string[], error: unknown): Error {
  const err = error as ExecError;
  const text = errorText(error);

  if (err.code === "ENOENT") {
    return new Error(
      `ADB executable not found: ${adbPath}. Configure Android SDK/ADB path in Settings.`
    );
  }

  if (/unauthorized/i.test(text)) {
    return new Error(
      "ADB device is unauthorized. Unlock the device and accept the USB debugging prompt, then retry."
    );
  }

  if (/device offline/i.test(text)) {
    return new Error(
      "ADB device is offline. Reconnect the device (or restart the emulator) and retry."
    );
  }

  if (isTransientDaemonError(error)) {
    return new Error(
      "ADB daemon is unavailable. Please retry; if it continues, restart adb.exe or reboot your machine."
    );
  }

  const detail = text
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter(Boolean)
    .slice(0, 4)
    .join(" | ");

  return new Error(
    `ADB command failed (${args.join(" ")}). ${detail || err.message || "Unknown error"}`
  );
}

async function runAdbRaw(
  adbPath: string,
  args: string[],
  timeout: number = TIMEOUT
): Promise<{ stdout: string; stderr: string }> {
  const { stdout, stderr } = await execFileAsync(adbPath, args, {
    timeout,
    windowsHide: true,
    maxBuffer: MAX_BUFFER,
  });

  return {
    stdout: toText(stdout),
    stderr: toText(stderr),
  };
}

async function runAdbRawBuffer(
  adbPath: string,
  args: string[],
  timeout: number = TIMEOUT
): Promise<{ stdout: Buffer; stderr: string }> {
  const { stdout, stderr } = await execFileAsync(adbPath, args, {
    timeout,
    windowsHide: true,
    maxBuffer: MAX_BUFFER,
    encoding: "buffer" as unknown as string,
  });

  const outBuffer = Buffer.isBuffer(stdout)
    ? stdout
    : Buffer.from(toText(stdout), "utf-8");

  return {
    stdout: outBuffer,
    stderr: toText(stderr),
  };
}

async function tryStartServer(adbPath: string): Promise<void> {
  try {
    await runAdbRaw(adbPath, ["start-server"], 5000);
  } catch {
    // best effort
  }
}

export async function runAdbCommand(
  adbPath: string,
  args: string[],
  timeout: number = TIMEOUT
): Promise<{ stdout: string; stderr: string }> {
  try {
    return await runAdbRaw(adbPath, args, timeout);
  } catch (error) {
    if (isTransientDaemonError(error)) {
      await tryStartServer(adbPath);
      await sleep(300);
      try {
        return await runAdbRaw(adbPath, args, timeout);
      } catch (retryError) {
        throw formatAdbError(adbPath, args, retryError);
      }
    }
    throw formatAdbError(adbPath, args, error);
  }
}

async function runAdbBufferCommand(
  adbPath: string,
  args: string[],
  timeout: number = TIMEOUT
): Promise<{ stdout: Buffer; stderr: string }> {
  try {
    return await runAdbRawBuffer(adbPath, args, timeout);
  } catch (error) {
    if (isTransientDaemonError(error)) {
      await tryStartServer(adbPath);
      await sleep(300);
      try {
        return await runAdbRawBuffer(adbPath, args, timeout);
      } catch (retryError) {
        throw formatAdbError(adbPath, args, retryError);
      }
    }
    throw formatAdbError(adbPath, args, error);
  }
}

export interface AdbRepairResult {
  success: boolean;
  adbPath: string;
  message: string;
  details: string[];
  devicesFound: number;
}

export interface AdbDiagnosticsResult {
  adbPath: string;
  version: string;
  devicesFound: number;
  details: string[];
  rawDevices: string[];
}

export async function repairAdb(adbPath: string): Promise<AdbRepairResult> {
  const details: string[] = [];

  try {
    await runAdbRaw(adbPath, ["kill-server"], 5000);
    details.push("Stopped existing ADB server process.");
  } catch (error) {
    const err = error as ExecError;
    if (err.code === "ENOENT") {
      throw formatAdbError(adbPath, ["kill-server"], error);
    }
    details.push("ADB server was not running or could not be stopped cleanly.");
  }

  await sleep(250);

  try {
    await runAdbRaw(adbPath, ["start-server"], 7000);
    details.push("Started ADB server.");
  } catch (error) {
    throw formatAdbError(adbPath, ["start-server"], error);
  }

  await sleep(300);

  const { stdout } = await runAdbCommand(adbPath, ["devices", "-l"]);
  const rawDevices = stdout
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("List of"));

  const onlineDevices = rawDevices.filter((line) => /\sdevice(\s|$)/.test(line)).length;
  details.push(
    `Detected ${rawDevices.length} attached device(s), ${onlineDevices} online.`
  );

  return {
    success: true,
    adbPath,
    message: `ADB repair completed successfully. Found ${rawDevices.length} device(s).`,
    details,
    devicesFound: rawDevices.length,
  };
}

export async function getAdbDiagnostics(adbPath: string): Promise<AdbDiagnosticsResult> {
  const details: string[] = [];

  const { stdout: versionOut } = await runAdbCommand(adbPath, ["version"], 5000);
  const versionLine = versionOut
    .split(/\r?\n/)
    .map((line) => line.trim())
    .find((line) => /^android debug bridge version/i.test(line));
  const version = versionLine || versionOut.trim() || "Unknown";
  details.push(`ADB version: ${version}`);

  const { stdout: devicesOut } = await runAdbCommand(adbPath, ["devices", "-l"]);
  const rawDevices = devicesOut
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("List of"));

  const onlineDevices = rawDevices.filter((line) => /\sdevice(\s|$)/.test(line)).length;
  const unauthorized = rawDevices.filter((line) => /\sunauthorized(\s|$)/.test(line)).length;
  const offline = rawDevices.filter((line) => /\soffline(\s|$)/.test(line)).length;

  details.push(`Detected devices: ${rawDevices.length}`);
  details.push(`Online: ${onlineDevices}`);
  if (unauthorized > 0) details.push(`Unauthorized: ${unauthorized}`);
  if (offline > 0) details.push(`Offline: ${offline}`);
  if (rawDevices.length === 0) details.push("No connected devices or emulators detected.");

  return {
    adbPath,
    version,
    devicesFound: rawDevices.length,
    details,
    rawDevices,
  };
}

// ---------------------------------------------------------------------------
// Device listing
// ---------------------------------------------------------------------------

export async function listDevices(adbPath: string): Promise<AdbDevice[]> {
  const { stdout } = await runAdbCommand(adbPath, ["devices", "-l"]);

  const rawDevices: Array<{
    serial: string;
    status: string;
    model: string;
    product: string;
    transportId: string;
    isEmulator: boolean;
    deviceType: string;
  }> = [];

  for (const line of stdout.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("List of")) continue;

    const parts = trimmed.split(/\s+/);
    if (parts.length < 2) continue;

    const serial = parts[0];
    const status = parts[1];
    let model = "";
    let product = "";
    let transportId = "";

    for (const part of parts.slice(2)) {
      if (part.startsWith("model:")) model = part.slice(6);
      else if (part.startsWith("product:")) product = part.slice(8);
      else if (part.startsWith("transport_id:")) transportId = part.slice(13);
    }

    const isEmulator =
      serial.startsWith("emulator-") || serial.startsWith("localhost:");
    const deviceType = status === "device" ? (isEmulator ? "emulator" : "physical") : status;

    rawDevices.push({ serial, status, model, product, transportId, isEmulator, deviceType });
  }

  // Enrich online devices with Android version + API level (parallel)
  const devices: AdbDevice[] = await Promise.all(
    rawDevices.map(async (d) => {
      let androidVersion = "";
      let apiLevel = 0;
      if (d.status === "device") {
        const [ver, sdk] = await Promise.all([
          getProp(adbPath, d.serial, "ro.build.version.release"),
          getProp(adbPath, d.serial, "ro.build.version.sdk"),
        ]);
        androidVersion = ver;
        apiLevel = parseInt(sdk, 10) || 0;
      }
      return { ...d, androidVersion, apiLevel };
    })
  );

  return devices;
}

// ---------------------------------------------------------------------------
// Device details
// ---------------------------------------------------------------------------

async function getProp(
  adbPath: string,
  serial: string,
  prop: string
): Promise<string> {
  try {
    const { stdout } = await runAdbCommand(adbPath, [
      "-s",
      serial,
      "shell",
      "getprop",
      prop,
    ]);
    return stdout.trim();
  } catch {
    return "";
  }
}

export async function getDeviceDetails(
  adbPath: string,
  serial: string
): Promise<AdbDeviceDetails> {
  const [
    manufacturer,
    model,
    androidVersion,
    sdkVersion,
    buildNumber,
    hardware,
    abi,
    locale,
    timezone,
    networkOperator,
    screenSize,
    batteryLevel,
  ] = await Promise.all([
    getProp(adbPath, serial, "ro.product.manufacturer"),
    getProp(adbPath, serial, "ro.product.model"),
    getProp(adbPath, serial, "ro.build.version.release"),
    getProp(adbPath, serial, "ro.build.version.sdk"),
    getProp(adbPath, serial, "ro.build.display.id"),
    getProp(adbPath, serial, "ro.hardware"),
    getProp(adbPath, serial, "ro.product.cpu.abi"),
    getProp(adbPath, serial, "persist.sys.locale"),
    getProp(adbPath, serial, "persist.sys.timezone"),
    getProp(adbPath, serial, "gsm.operator.alpha"),
    runAdbCommand(adbPath, ["-s", serial, "shell", "wm", "size"])
      .then(({ stdout }) => {
        const match = stdout.match(/Physical size:\s*(\d+x\d+)/);
        return match ? match[1] : "";
      })
      .catch(() => ""),
    runAdbCommand(adbPath, ["-s", serial, "shell", "dumpsys", "battery"])
      .then(({ stdout }) => {
        const match = stdout.match(/level:\s*(\d+)/);
        return match ? match[1] : "";
      })
      .catch(() => ""),
  ]);

  return {
    serial,
    manufacturer,
    model,
    androidVersion,
    sdkVersion: parseInt(sdkVersion, 10) || 0,
    buildNumber,
    hardware,
    abi,
    locale,
    timezone,
    networkOperator,
    screenSize,
    batteryLevel: parseInt(batteryLevel, 10) || 0,
  };
}

// ---------------------------------------------------------------------------
// Screenshot
// ---------------------------------------------------------------------------

export async function takeScreenshot(
  adbPath: string,
  serial: string
): Promise<string> {
  const { stdout } = await runAdbBufferCommand(adbPath, [
    "-s",
    serial,
    "exec-out",
    "screencap",
    "-p",
  ]);
  return stdout.toString("base64");
}

// ---------------------------------------------------------------------------
// Device input
// ---------------------------------------------------------------------------

export async function deviceInput(
  adbPath: string,
  serial: string,
  action: string,
  x?: number,
  y?: number,
  x2?: number,
  y2?: number,
  duration?: number,
  text?: string,
  keyCode?: number
): Promise<string> {
  let args: string[];

  switch (action) {
    case "tap":
      if (x == null || y == null) throw new Error("tap requires x and y");
      args = ["-s", serial, "shell", "input", "tap", `${x}`, `${y}`];
      break;
    case "swipe":
      if (x == null || y == null || x2 == null || y2 == null)
        throw new Error("swipe requires x, y, x2, y2");
      args = [
        "-s", serial, "shell", "input", "swipe",
        `${x}`, `${y}`, `${x2}`, `${y2}`, `${duration ?? 300}`,
      ];
      break;
    case "longPress":
      if (x == null || y == null) throw new Error("longPress requires x and y");
      args = [
        "-s", serial, "shell", "input", "swipe",
        `${x}`, `${y}`, `${x}`, `${y}`, `${duration ?? 1000}`,
      ];
      break;
    case "inputText":
      if (!text) throw new Error("inputText requires text");
      args = ["-s", serial, "shell", "input", "text", text.replace(/ /g, "%s")];
      break;
    case "keyEvent":
      if (keyCode == null) throw new Error("keyEvent requires keyCode");
      args = ["-s", serial, "shell", "input", "keyevent", `${keyCode}`];
      break;
    case "dumpUi":
      await runAdbCommand(adbPath, [
        "-s",
        serial,
        "shell",
        "uiautomator",
        "dump",
        "/sdcard/window_dump.xml",
      ]);
      const { stdout: xmlOut } = await runAdbCommand(adbPath, [
        "-s",
        serial,
        "shell",
        "cat",
        "/sdcard/window_dump.xml",
      ]);
      return xmlOut;
    default:
      throw new Error(`Unknown action: ${action}`);
  }

  await runAdbCommand(adbPath, args);
  return "ok";
}

// ---------------------------------------------------------------------------
// APK install / uninstall
// ---------------------------------------------------------------------------

export async function installApk(
  adbPath: string,
  serial: string,
  apkPath: string,
  reinstall: boolean
): Promise<string> {
  const args = ["-s", serial, "install"];
  if (reinstall) args.push("-r");
  args.push("-g", apkPath);

  const { stdout } = await runAdbCommand(adbPath, args, 120_000);
  return stdout.trim();
}

export async function uninstallPackage(
  adbPath: string,
  serial: string,
  packageName: string
): Promise<string> {
  const { stdout } = await runAdbCommand(
    adbPath,
    ["-s", serial, "uninstall", packageName],
    30_000
  );
  return stdout.trim();
}

// ---------------------------------------------------------------------------
// Packages
// ---------------------------------------------------------------------------

export async function listPackages(
  adbPath: string,
  serial: string
): Promise<string[]> {
  const { stdout } = await runAdbCommand(adbPath, [
    "-s",
    serial,
    "shell",
    "pm",
    "list",
    "packages",
    "-3",
  ]);
  return stdout
    .split("\n")
    .map((l) => l.trim().replace(/^package:/, ""))
    .filter(Boolean);
}

// ---------------------------------------------------------------------------
// Logcat
// ---------------------------------------------------------------------------

export async function getLogcat(
  adbPath: string,
  serial: string,
  lines: number
): Promise<string[]> {
  const { stdout } = await runAdbCommand(adbPath, [
    "-s",
    serial,
    "logcat",
    "-d",
    "-t",
    `${lines}`,
    "-v",
    "threadtime",
  ]);
  return stdout.split("\n").filter(Boolean);
}

export async function clearLogcat(
  adbPath: string,
  serial: string
): Promise<string> {
  await runAdbCommand(adbPath, ["-s", serial, "logcat", "-c"]);
  return "ok";
}
